### DISCUSSION TITLE
Docker image security and updates for older keycloak versions

### DISCUSSION THREAD

**Discussion by XcrigX** [keycloak/keycloak#23176] [Q&A]
I'm trying to determine if there is a better way to keep my keycloak images up to date. 

Here's the basic issue - Keycloak has undergone several major version revisions the past couple of years. I have custom extensions built, as well as CLI configurers, etc. which are likely sensitive to the underlying keycloak version. 
Hence, when I build my docker image to deploy, I specify a specific keycloak version. 
i.e. instead of: 
FROM quay.io/keycloak/keycloak:latest
I'm using
FROM quay.io/keycloak/keycloak:20.0.3 (or whatever version). 

But when using tagged versions like this which are static, security vulnerabilities and CVEs in the base images/OS are never updated. 

It would be ideal if there were specific per-keycloak-version images which were occasionally released. 
i.e. something like:
FROM quay.io/keycloak/keycloak-19:latest
FROM quay.io/keycloak/keycloak-20:latest
FROM quay.io/keycloak/keycloak-21:latest
...

Anyway, I guess this is part suggestion, part question.   
But the question is, how are other people dealing with this?  Or, any suggestions for how someone could go about getting the images up to date without updating the keycloak version itself?

**Comment by DAHAG-ArisNourbakhsh** [comment]
We use renovatebot https://docs.renovatebot.com/modules/datasource/docker/
Which is essentially "updating the keycloak version itself" but mostly automated.

**Comment by XcrigX** [comment]
NOTE: I found a related, but likely unresolved discussion here: https://github.com/keycloak/keycloak/discussions/12082

**Comment by vmuzikar** [comment]
We're providing floating tags that locks the version to a minor release, e.g. `quay.io/keycloak/keycloak:22.0`. This is updated every time we release a 22.0 micro, e.g. 22.0.1, 22.0.2, ...

Please note we always support only the latest release version, we do not provide security fixes for e.g. Keycloak 21.1 anymore.

**Reply by XcrigX** [reply]
Hi @vmuzikar - thanks for the response. I think that's a problem for downstream users. 
And I'm not talking about security fixes for Keycloak itself, I understand that.  

We regularly scan all running containers for active CVEs. A high percentage of detected CVEs are with the base images themselves - packages in the OS.  For most of the apps I maintain, I can simply rebuild my image - which pulls the latest base image, and problem solved. This can even be automated to rebuild and deploy on a schedule. 
Using the Keycloak images would require users to constantly check the latest version, check the change logs, check the migration steps if applicable, re-test everything.  With the breakneck speed at which Keycloak advanced from v19 to v22 with major API and UI changes - it's not very feasible for folks that are extending and customizing Keycloak. 

Anyway, the best solution I've come up with is to copy the Keycloak Dockerfile and build it myself so I can get UBI image updates whenever I need them without advancing the Keycloak version. 

Eventually, there would be a CVE detected in one of the java dependencies of Keycloak - forcing an version update, but in my experience these are less frequent than CVEs identified for linux packages and allows for a less frequent upgrade cycle.
